core/domain/topic_jobs_one_off.py

```diff
@@ -92,3 +94,65 @@ def reduce(key, values):
                 sum(ast.literal_eval(v) for v in values))])
         else:
             yield (key, values)
+
+
+class RemoveDeletedSkillsFromTopicOneOffJob(
+        jobs.BaseMapReduceOneOffJobManager):
+    """One-off job to remove deleted uncategorized skills linked to a topic."""
+
+    _DELETED_KEY = 'topic_deleted'
+
+    @classmethod
+    def entity_classes_to_map_over(cls):
+        return [topic_models.TopicModel]
+
+    @staticmethod
+    def map(item):
+        if item.deleted:
+            yield (RemoveDeletedSkillsFromTopicOneOffJob._DELETED_KEY, 1)
+            return
+
+        # Note: the read will bring the topic up to the newest version.
+        topic = topic_fetchers.get_topic_by_id(item.id)
+        skill_ids_to_be_removed_from_subtopic = []
+        all_skill_ids_to_be_removed = []
+        commit_cmds = []
+        for subtopic in topic.get_all_subtopics():
```

Comment:
Maybe add brief comments about this block: "This block of code removes deleted skills from subtopics, but keeps them in the topic"; "This block of code removes all deleted skills from topics".

In general, if a reviewer asks about something, always strongly consider writing an explanatory comment in the code itself, since it typically means that something about the existing code is not obvious to a fellow developer. Otherwise the next request from the reviewer will be "please add comments to the code". You can often save a round of review that way.

Also, could you please check that the prod validation jobs ensure that this is caught? I.e. that it is a requirement that neither a subtopic nor a topic references any deleted skills? If that's not checked there then we need to add that.

Reply:
> Maybe add brief comments about this block: "This block of code removes deleted skills from subtopics, but keeps them in the topic"; "This block of code removes all deleted skills from topics".

Done.

> In general, if a reviewer asks about something, always strongly consider writing an explanatory comment in the code itself, since it typically means that something about the existing code is not obvious to a fellow developer. Otherwise the next request from the reviewer will be "please add comments to the code". You can often save a round of review that way.

Yes, noted. In this case though, I wanted to clarify/discuss my approach here and capture the final decision in the comment (ie. it could be that, in the course of the discussion, we realise that my approach is incorrect and I have to change it). But I now realise that if I address a comment without a code change, it makes sense to add a comment to the code so that it avoids an extra round of review (if my approach is indeed correct.)

> Also, could you please check that the prod validation jobs ensure that this is caught? I.e. that it is a requirement that neither a subtopic nor a topic references any deleted skills? If that's not checked there then we need to add that.

Looks like prod_validation_jobs_one_off.py already does this using `_validate_external_id_relationships` which runs for all model validations (ie. ensures external id properties on the model correspond to valid instances.) So no changes are required there.

Reply:
> Yes, noted. In this case though, I wanted to clarify/discuss my approach here and capture the final decision in the comment (ie. it could be that, in the course of the discussion, we realise that my approach is incorrect and I have to change it). But I now realise that if I address a comment without a code change, it makes sense to add a comment to the code so that it avoids an extra round of review (if my approach is indeed correct.)

Yup -- the thing to do is, always add a comment to reflect your current understanding (no harm doing so). As your understanding evolves, keep the comment updated accordingly, just as you might change the code accordingly as the requirements become clearer.